[Scrollable] refactor: rewrite Scrollable utilizing platform features

iAmNathanJ opened the pull request:
<!--
  ☝️How to write a good PR title:
  - Prefix it with [ComponentName] (if applicable), for example: [Button]
  - Start with a verb, for example: Add, Delete, Improve, Fix…
  - Give as much context as necessary and as little as possible
  - Prefix it with [WIP] while it’s a work in progress
-->

### WHY are these changes introduced?

This updates the Scrollable component to be a function component, but also changes the fundamental implementation strategy.

It no longer controls the DOM node's scroll, allowing the platform to handle all scrolling interaction and only using native features to supply enhancements for the `ScrollTo` component and the scroll `hint` functionality.

Primary Benefits:
- It allows for smooth scrolling in React 18 on iOS. Since we are no longer setting scroll position on the target element as a side effect of state, it has no conflict with automatic batching by React. See https://github.com/Shopify/polaris/issues/7424
- There are minimal wasted renders when scrolling. Previously, we would render many times just by virtue of scrolling.

resolves #7424

### WHAT is this pull request doing?

<details>
<summary>Before/After :eyes:</summary>

#### Before

![](https://screenshot.click/19-54-tmbl7-1k4ao.png)

Momentum scrolling interrupted by render/commit cycles

https://user-images.githubusercontent.com/8847861/196776562-0517c364-9992-4415-973d-6d77843a4f0b.mov

#### After

![](https://screenshot.click/19-04-seyci-ivgq1.png)

Smooth scrolling

https://user-images.githubusercontent.com/8847861/196776608-5464f140-ee9b-437b-9229-6a80f4cb7a87.mov

</details>

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

Access storybook on an iOS device:
![](https://screenshot.click/19-38-t6tq5-tgs56.png)

Using iOS device, navigate to the `Scrollable` component:
`http://<your.network.ip>:6006/?path=/story/all-components-scrollable--default&globals=profiler:true`

- [ ] On `main` branch: scrolling should be janky - experience is choppy, momentum gets interrupted
- [ ] On this branch: scrolling should be smooth
  - [ ] `shadow` prop works as intended
  - [ ] `hint` prop works as intended
  - [ ] `ScrollTo` component works within `Scrollable`

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

iAmNathanJ committed:
refactor: rewrite Scrollable utilizing platform features

This updates the Scrollable component to be a function component, but
also changes the fundamental implementation strategy.

Previously, the Scrollable component held internal state for the scroll
position and controlled the scroll of the actual DOM node as a side
effect.

With this change, there is no controlling of the DOM node's scroll,
allowing the platform to handle all scrolling interaction and only using
native features to supply enhancements for the ScrollTo component, the
hint functionality.

iAmNathanJ committed:
add story for Scrollable using the hint prop

kyledurand commented on .changeset/many-otters-repeat.md:
```diff
@@ -0,0 +1,5 @@
+'@shopify/polaris': patch
+---
+
+Improve performance of the Scrollable component with React 18
```
```suggestion
Converted `Scrollable` to functional component and improve its performance with React 18
```

kyledurand commented on polaris-react/src/components/Scrollable/Scrollable.tsx:
```diff
@@ -36,218 +32,86 @@ export interface ScrollableProps extends React.HTMLProps<HTMLDivElement> {
+  const stickyManager = useRef(new StickyManager());
+  const scrollArea = useRef<HTMLDivElement>(null);
+  const scrollTo = useCallback((scrollY: number) => {
+    scrollArea.current?.scrollTo({top: scrollY, behavior: 'smooth'});
```
Probably out of scope for this PR but I'd love if we could add a check for perfers reduced motion and only use smooth behaviour if that check returns false

If you have time it would be a nice add, I think

iAmNathanJ replied:
Had the same thought! Thanks for the callout, I'll update.

AndrewMusgrave replied:
It looks like the component previous had the check - we may want to include it in this pull 😄

kyledurand commented on polaris-react/src/components/Scrollable/Scrollable.tsx:
```diff
@@ -36,218 +32,86 @@ export interface ScrollableProps extends React.HTMLProps<HTMLDivElement> {
+      currentScrollArea.removeEventListener('scroll', handleScroll);
+      window.removeEventListener('resize', handleResize);
+    };
+  }, [shadow]);
```
My only minor concern with the code in this PR is that the `shadow` work is encapsulated within a useEffect when it seems like it shouldn't have to be. It's probably fine so not a blocker, it's kind of nice this way anyway but might be causing an unnecessary render, sometimes.

iAmNathanJ replied:
Oh interesting. I think setting the state properties needs to happen in the effect since it's the only responsibility of the scroll handler? But the `shadow` prop doesn't need to be part of that and could instead [move down as an added condition for the classNames](https://github.com/Shopify/polaris/pull/7443/commits/19af0a05030c483c5b0b908e2c7295db16481934#diff-5e25b3e154b4371b6b411b066a493084f30d02dabad5345a00906f95ce133c3fR95-R96). Is that what you're saying? I think I'll do that so we can remove it from the dependency array here. But let me know if I'm misunderstanding something.

kyledurand reviewed (commented):
This looks fantastic but I haven't had a chance to finish tophatting in the admin yet. I'm also trying to get a couple more folks with context to take a look in the meantime

AndrewMusgrave commented on polaris-react/src/components/Scrollable/Scrollable.tsx:
```diff
@@ -36,218 +32,86 @@ export interface ScrollableProps extends React.HTMLProps<HTMLDivElement> {
+}: ScrollableProps) {
+  const [topShadow, setTopShadow] = useState(false);
+  const [bottomShadow, setBottomShadow] = useState(false);
+  const stickyManager = useRef(new StickyManager());
```
Does sticky manager have a lot of over head in the constructor? We may want to use a lazy ref here. Even though the instance of sticky manager never changes, a new one will be created every render.

iAmNathanJ replied:
Great callout. I don't think there's much overhead since we don't pass the container here, but I'd be happy to update with lazy ref.

iAmNathanJ replied:
Unfortunately this forces me to put `stickyManager` into the dependency array on line 88. Thoughts on disabling the lint rule for that? I'd probably prefer to just leave it in the array since its reference shouldn't change and won't actually trigger extra effects.

AndrewMusgrave commented on polaris-react/src/components/Scrollable/Scrollable.tsx:
```diff
@@ -36,218 +32,86 @@ export interface ScrollableProps extends React.HTMLProps<HTMLDivElement> {
+  }, []);
+
+  useEffect(() => {
+    if (hint) {
```
Not sure if we care, but this is slightly different than the previous implementation. The hint used to only be on mount

iAmNathanJ replied:
I considered this as well. Opted to change because:
- from a maintenance perspective, trying to avoid the cost of adding something like `useEffectOnce` (to mimic componentDidMount) or disabling exhaustive deps
- it could be desired behavior for a consumer to toggle the prop and trigger the hint sometime after initial render

That said, it's not 1:1 with the old. It's probably more responsible to make it 1:1 (as much as possible) unless we want to push the new version higher than patch.

iAmNathanJ replied:
Saw that we already have `useComponentDidMount` so going with that.

AndrewMusgrave replied:
> That said, it's not 1:1 with the old. It's probably more responsible to make it 1:1 (as much as possible) unless we want to push the new version higher than patch.

That's fair. I'm fine with either. It does seem a little odd to only use the value on mount, since that's not reactive. But it was the previous implementation 🤷

AndrewMusgrave commented on polaris-react/src/components/Scrollable/Scrollable.tsx:
```diff
@@ -36,218 +32,86 @@ export interface ScrollableProps extends React.HTMLProps<HTMLDivElement> {
+  shadow,
+  hint,
+  focusable,
+  onScrolledToBottom,
```
I might be missing it - are we still using this prop? It looks like it used to be called in `handleScroll` before

iAmNathanJ replied:
Nope, I missed this one. Thank you, I think it needs to stay and be used. I'll fix it. Admittedly, I was rushing through this a bit trying to get feedback on the fundamental approach before fully refining.

AndrewMusgrave commented on polaris-react/src/components/Scrollable/Scrollable.tsx:
```diff
@@ -36,218 +32,86 @@ export interface ScrollableProps extends React.HTMLProps<HTMLDivElement> {
-      horizontal && styles.horizontal,
-      topShadow && styles.hasTopShadow,
-      bottomShadow && styles.hasBottomShadow,
-      vertical && canScroll && styles.verticalHasScrolling,
```
Was deleting this intentional & if so, should we remove it from the scss file?

iAmNathanJ replied:
Good eye, I was thinking it should come out of the css. It wasn't clear to me how this is needed when scroll container already has overflow auto. Do you have context on why the class was there?

AndrewMusgrave replied:
Nope, no context and you're right. The `verticalHasScrolling` class already applies the overflow. This likely isn't needed

AndrewMusgrave commented on polaris-react/src/components/Scrollable/Scrollable.tsx:
```diff
@@ -36,218 +32,86 @@ export interface ScrollableProps extends React.HTMLProps<HTMLDivElement> {
-      return;
-    }
+    stickyManager.current?.setContainer(currentScrollArea);
+    currentScrollArea.addEventListener('scroll', handleScroll);
```
Should we keep this within RFA to avoid reflows?

iAmNathanJ replied:
I was wrestling with this question myself thinking react scheduler should be responsible for managing DOM writes, but I think even reading scroll position in `handleScroll` triggers reflow. However, I'm not seeing the difference in performance profiles (chrome) when adding rAF. I was also considering if basic throttling might be a good option. Maybe you know better, are we sure there's a meaningful difference to use rAF in this case?

iAmNathanJ replied:
Ok, I captured some rendering profiles using a **real iPhone** this time. The two seem to be more or less identical. The interaction I tested was scrolling to bottom of the container and back up - using the default story in storybook. Again, let me know if you feel strongly or know for sure we need to use `rAF`.

**with requestAnimationFrame**
![](https://screenshot.click/25-12-g71u9-k9vk7.png)

**without requestAnimationFrame**
![](https://screenshot.click/25-14-9htd7-c5d7s.png)

iAmNathanJ replied:
Profiling output aside, I went ahead and pushed a commit to add requestAnimationFrame https://github.com/Shopify/polaris/pull/7443/commits/bf52e2e78111f4f285291e7796f44503e9bf61cb

AndrewMusgrave replied:
Thanks for making the change ❤️ Tbh I'm not sure if it'll have that much of an impact, but we have seen issues on larger pages. [This pull](https://github.com/Shopify/polaris/pull/6840) from a few months ago saved a decent chunk of time taking advantage of RaF

AndrewMusgrave commented on polaris-react/src/components/Scrollable/Scrollable.tsx:
```diff
@@ -36,218 +32,86 @@ export interface ScrollableProps extends React.HTMLProps<HTMLDivElement> {
+
+  useEffect(() => {
+    if (hint) {
+      performScrollHint(scrollArea.current);
```
^ (comment re: RFA)

AndrewMusgrave reviewed (commented):
Looking good 🙌 Left a few comments to gain some context!

iAmNathanJ committed:
Address PR Feedback

- remove unneeded css class
- respect reduced motion settings in scrollTo function
- use lazy ref for sticky manager
- fix: scroll hint happens at max once per mount
- fix: invoke callback when scrolled to bottom of container

iAmNathanJ committed:
add story for onScrollToBottom scenario

tmlayton commented on polaris-react/src/components/Scrollable/Scrollable.stories.tsx:
```diff
@@ -416,3 +446,84 @@ export function ToChildComponent() {
+          update and change the Terms of Service by posting updates and changes
+          to the Shopify website. You are advised to check the Terms of Service
+          from time to time for any updates or changes that may impact you.
+        </p>
```
Could you make this copy longer so it’s easier to test scrolling?

iAmNathanJ replied:
Done! https://github.com/Shopify/polaris/pull/7443/commits/63c2b5b4358031e34072f260ed2e6b8bea82af77

tmlayton commented on polaris-react/src/components/Scrollable/Scrollable.tsx:
```diff
@@ -40,218 +38,93 @@ export interface ScrollableProps extends React.HTMLProps<HTMLDivElement> {
-    });
-  };
+      if (isAtBottomOfScroll && onScrolledToBottom) {
+        onScrolledToBottom();
```
This appears to be preexisting, but the scroll-to-bottom callback will get called 2-3 times when scrolling to the bottom and 2-3 more times when starting from the bottom and scrolling up.

iAmNathanJ replied:
I noticed this too and checked with the old implementation. I suspect we'd want to keep it this way as changing could break some consumer. But agree this would ideally be called only once (maybe with retry capability).

iAmNathanJ replied:
Following up for posterity, I believe this happens because we receive multiple scroll events within the "buffer zone" for the callback. Original PR that introduced that for reference https://github.com/Shopify/polaris/pull/4218.

tmlayton reviewed (approved):
Couple minor comments, otherwise 👍

iAmNathanJ committed:
use requestAnimationFrame to avoid potential reflows in scroll event handlers

AndrewMusgrave commented on polaris-react/src/components/Scrollable/Scrollable.tsx:
```diff
@@ -40,218 +38,95 @@ export interface ScrollableProps extends React.HTMLProps<HTMLDivElement> {
+      currentScrollArea.removeEventListener('scroll', handleScroll);
+      globalThis.removeEventListener('resize', handleResize);
+    };
+  }, [stickyManager, onScrolledToBottom]);
```
One concern I have is that if `onScrolledToBottom` isn't memoized and the parent re-renders frequently on scroll, we'll be polluted with event listeners and `onScrolledToBottom` will be called multiple times as each debounce will be unique. Something we'll want to keep an eye on when rolling this out 👀

iAmNathanJ replied:
Great eye. I wonder if this could trigger cyclical renders. 🤔 I'm gonna look at this a little closer before shipping.

iAmNathanJ replied:
resolved https://github.com/Shopify/polaris/pull/7443/commits/cd7fc7fb1b690616bb66714a698a3b4ead14d1c1

AndrewMusgrave reviewed (approved):
Nice work, LGTM 🙌  Left a comment re: a dependency on use effect that we'll want to keep an eye on!

iAmNathanJ committed:
fix: infinite render cycles caused by onScrolledToBottom prop

kyledurand reviewed (commented):
👏 Code looks good. Haven't 🎩 'd the latest snapshot though

kyledurand commented on .changeset/many-otters-repeat.md:
```diff
@@ -0,0 +1,5 @@
+'@shopify/polaris': patch
+---
+
+Improve performance of the Scrollable component with React 18
```
We try to keep entries in past tense

```suggestion
Improved performance of the Scrollable component with React 18
```

iAmNathanJ replied:
wowzers, I missed this comment before I merged. Sorry about that!
